resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/KubernetesExternalShuffleService.scala

```diff
@@ -0,0 +1,133 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.deploy.kubernetes
+
+import java.nio.ByteBuffer
+
+import scala.collection.JavaConverters._
+
+import io.fabric8.kubernetes.api.model.Pod
+import io.fabric8.kubernetes.client.{KubernetesClientException, Watch, Watcher}
+import io.fabric8.kubernetes.client.Watcher.Action
+
+import org.apache.spark.{SecurityManager, SparkConf}
+import org.apache.spark.deploy.ExternalShuffleService
+import org.apache.spark.deploy.kubernetes.constants._
+import org.apache.spark.internal.Logging
+import org.apache.spark.network.client.{RpcResponseCallback, TransportClient}
+import org.apache.spark.network.shuffle.ExternalShuffleBlockHandler
+import org.apache.spark.network.shuffle.protocol.BlockTransferMessage
+import org.apache.spark.network.shuffle.protocol.kubernetes.RegisterKubernetesApp
+import org.apache.spark.network.util.TransportConf
+import org.apache.spark.scheduler.cluster.kubernetes.DriverPodKubernetesClientProvider
+
+
+/**
+ * An RPC endpoint that receives registration requests from Spark drivers running on Kubernetes.
+ * It detects driver termination and calls the cleanup callback to [[ExternalShuffleService]].
+ */
+private[spark] class KubernetesShuffleBlockHandler (
+  sparkConf: SparkConf,
+  transportConf: TransportConf)
+  extends ExternalShuffleBlockHandler(transportConf, null) with Logging {
+
+  private var connectedApps: Set[String] = Set()
+  private val CONNECTED_APPS_LOCK = new Object
+  private val Watch: Option[Watch] = ShuffleWatcher()
+
+  protected override def handleMessage(
+    message: BlockTransferMessage,
+    client: TransportClient,
+    callback: RpcResponseCallback): Unit = {
+      message match {
+        case RegisterDriverParam(appId) =>
+          val address = client.getSocketAddress
+          logInfo(s"Received registration request from app $appId (remote address $address).")
+          CONNECTED_APPS_LOCK.synchronized {
+            if (connectedApps.contains(appId)) {
+              logWarning(s"Received a registration request from app $appId, but it was already " +
+                s"registered")
+            }
+            connectedApps += appId
+          }
+          callback.onSuccess(ByteBuffer.allocate(0))
+        case _ => super.handleMessage(message, client, callback)
+      }
+  }
+
+  private def ShuffleWatcher(): Option[Watch] = {
+    try {
+      val kubernetesClient = new DriverPodKubernetesClientProvider(sparkConf, "default").get
```

Comment:
Do we always anticipate the shuffle service running in the default namespace?

We should probably inject the Kubernetes client or the Kubernetes client provider in the constructor.